R/config.R

```diff
@@ -0,0 +1,250 @@
+#' @param exclusions An instance of `Exclusions` class containing exclusion
+#' criteria.
+#' @param config_version A numeric value specifying the configuration version.
+#' @param quantile_width A vector of numeric values representing the desired
+#' quantiles.
+#' @param model A string specifying the model to be used.
+#' @param report_date A string representing the report date. Formatted as
+#' "YYYY-MM-DD".
+#' @param as_of_date A string representing the as-of date. Formatted as
+#' "YYYY-MM-DD".
+#' @export
+Config <- S7::new_class(
+  "Config",
+  properties = list(
+    job_id = S7::class_character,
+    task_id = S7::class_character,
+    min_reference_date = S7::class_character,
+    max_reference_date = S7::class_character,
+    report_date = S7::class_character,
+    as_of_date = S7::class_character,
+    disease = S7::class_character,
+    geo_value = S7::class_character,
+    geo_type = S7::class_character,
+    seed = S7::class_integer,
+    horizon = S7::class_integer,
+    model = S7::new_property(S7::class_character, default = "EpiNow2"),
+    config_version = S7::class_character,
+    quantile_width = S7::new_property(S7::class_vector, default = c(0.5, 0.95)),
+    data = S7::S7_class(Data()),
+    # Using a list instead of an S7 object, because EpiNow2 expects a list, and
+    # because it reduces changes to the pipeline code.
+    # Adding a default that shows the required keys, with the expected types as
+    # the values. Should fail loudly if the default values are used, but will
+    # be useful to the user to see what is expected.
+    priors = S7::new_property(S7::class_list, default = list(
+      rt = list(mean = S7::class_numeric, sd = S7::class_numeric),
+      gp = list(alpha_sd = S7::class_numeric)
+    )),
+    parameters = S7::S7_class(Parameters()),
+    # Using a list instead of an S7 object, because stan expects a list, and
+    # because it reduces changes to the pipeline code.
+    # Adding a default that shows the required keys, with the expected types as
+    # the values. Should fail loudly if the default values are used, but will
+    # be useful to the user to see what is expected.
+    # Using a list here also reduces changes to the pipeline code.
+    sampler_opts = S7::new_property(S7::class_list, default = list(
+      cores = S7::class_integer,
+      chains = S7::class_integer,
+      iter_warmup = S7::class_integer,
+      iter_sampling = S7::class_integer,
+      max_treedepth = S7::class_integer,
+      adapt_delta = S7::class_numeric
+    )),
+    exclusions = S7::S7_class(Exclusions())
+  )
+)
+
+#' Read JSON Configuration into Config Object
+#'
+#' Reads a JSON file from the specified path and converts it into a `Config`
+#' object.
+#'
+#' @param config_path A string specifying the path to the JSON configuration
+#' file.
+#' @return An instance of the `Config` class populated with the data from the
+#' JSON file.
+#' @export
+read_json_into_config <- function(config_path) {
```

Comment:
Should this be a class method?

Reply:
I thought about it, and that does seem like the "cleaner" solution from a packaging perspective: nice generic function, and all that. But then I read the docs on [generics and methods](https://rconsortium.github.io/S7/articles/generics-methods.html), and the section on [method registration in packages](https://rconsortium.github.io/S7/articles/packages.html), and thought that the package would be simpler if it is just a regular function.

This felt simpler in terms of the amount of code we have to maintain, but I'm open to making it a class method.

Reply:
That's convincing! Thanks for looking into it.